hax/helper/generate_sysconf.py

```diff
@@ -0,0 +1,347 @@
+    def generate_confd(self, svc_id: str, hax_ep: str, motr_confdir: str):
+        fid = Fid(ObjT.PROCESS.value, int(svc_id))
+        ep = self.provider.get_service_ep(svc_id)
+        filename = f'm0d-{fid}'
+        contents = (f"MOTR_M0D_EP='{ep}'\n"
+                    f"MOTR_HA_EP='{hax_ep}'\n"
+                    f"MOTR_PROCESS_FID='{fid}'\n"
+                    f"MOTR_CONF_XC='{motr_confdir}/confd.xc'\n")
+        self._write_file(self.sysconf_dir + filename, contents)
+
+    def generate_ios(self, svc_id: str, hax_ep: str):
+        fid = Fid(ObjT.PROCESS.value, int(svc_id))
+        ep = self.provider.get_service_ep(svc_id)
+        meta_data = self.provider.get_ios_meta_data(svc_id)
+        filename = f'm0d-{fid}'
+        contents = (f"MOTR_M0D_EP='{ep}'\n"
+                    f"MOTR_HA_EP='{hax_ep}'\n"
+                    f"MOTR_PROCESS_FID='{fid}'\n")
+        if meta_data:
+            contents += f'MOTR_BE_SEG_PATH={meta_data}\n'
+        self._write_file(self.sysconf_dir + filename, contents)
+
+    def generate_s3(self, svc_id: str, hax_ep: str, s3_port: int):
+        profile_fid = self.provider.get_profile_fid()
+        fid = Fid(ObjT.PROCESS.value, int(svc_id))
+        ep = self.provider.get_service_ep(svc_id)
+        filename = f's3server-{fid}'
+        contents = (f"MOTR_PROFILE_FID={profile_fid}\n"
+                    f"MOTR_S3SERVER_EP='{ep}'\n"
+                    f"MOTR_HA_EP='{hax_ep}'\n"
+                    f"MOTR_PROCESS_FID='{fid}'\n"
+                    f"MOTR_S3SERVER_PORT={s3_port}\n")
+        self._write_file(self.sysconf_dir + filename, contents)
+
+    def generate_sysconfig(self, sysconf_dir: str, motr_confdir: str):
+        s3_port = 28071
+        self.sysconf_dir = sysconf_dir + '/'
+        if not os.path.isdir(self.sysconf_dir):
+            raise FileNotFoundError(f'{self.sysconf_dir} does not exist.')
+
+        IDs = self.get_all_svc_ids()
+
+        hax_ep = self.provider.get_service_ep(IDs['HAX_ID'][0])
+        for x in IDs['CONFD_IDs']:
+            self.generate_confd(x, hax_ep, motr_confdir)
+
+        for x in IDs['IOS_IDs']:
+            self.generate_ios(x, hax_ep)
+
+        for x in IDs['S3_IDs']:
+            self.generate_s3(x, hax_ep, s3_port)
+            s3_port += 1
+
+    def append_svcs_to_file(self, svcs, conf_file):
+        with open(conf_file) as consul_conf_file:
+            data = simplejson.load(consul_conf_file)
+        data['services'] = svcs
+        self._write_file(conf_file,
+                         simplejson.dumps(data, indent=2, for_json=True))
+
+    def get_service_ipaddr(self, ep: str) -> str:
+        return ep.split('@')[0]
+
+    def append_ipaddr_to_file(self, hax_ep: str, conf_file: str):
+        ipaddr = self.get_service_ipaddr(hax_ep)
+        with fileinput.FileInput(conf_file, inplace=True,
+                                 backup='.bak') as file:
+            for line in file:
+                print(line.replace('http://localhost',
```

Comment:
It is just printing to the console, isn't it?

Why the function name is `append_ipaddr_to_file` then? I would expect that the function modifies the file somehow but it doesn't happen.

Reply:
Actually, FileInput supports in-place replacing- here it **redirects standard output to the input file** (first the input file is moved to the backup file), and it is deleted when the file is closed.
ref - https://docs.python.org/3/library/fileinput.html#fileinput.FileInput

Reply:
It does, but I guess that `file` should be passed as a [parameter to print](https://stackoverflow.com/questions/36571560/directing-print-output-to-a-txt-file) then. Am I wrong?

Reply:
[in-place editing file](https://www.geeksforgeeks.org/python-inplace-editing-using-fileinput/ ) here, stdout is redirected to file within that context manager.

Reply:
And I tested the code.. it seems to work as expected.

Reply:
I suspected something like that but I didn't manage to find a proof in the official documents. If you have the link that describes the behaviour of context manager + print, please add to the code as a comment. Otherwise it looks too magic.

Reply:
> Optional in-place filtering: if the keyword argument inplace=True is passed to the FileInput constructor, the file is moved to a backup file and standard output is directed to the input file. This makes it possible to write a filter that rewrites its input file in place. 

Basically, the file taken in constructor of FileInput .. the stdout is directed to that file.

Reply:
Done added the link and explanation as comments.